algorithms/linfa-ftrl/src/algorithm.rs

```diff
@@ -0,0 +1,346 @@
+        Array1::zeros(x.nrows()).mapv(Pr::new)
+    }
+}
+
+/// View the fitted parameters and make predictions with a fitted
+/// follow the regularized leader -proximal, model
+impl<F: Float> FTRL<F> {
+    /// Get Z values
+    pub fn z(&self) -> &Array1<F> {
+        &self.z
+    }
+
+    /// Get N values
+    pub fn n(&self) -> &Array1<F> {
+        &self.n
+    }
+
+    /// Get the hyperparameters
+    pub fn get_params(&self) -> &FtrlValidParams<F> {
+        &self.params
+    }
+
+    /// Calculate weights for model prediction
+    pub fn get_weights(&self) -> Array1<F> {
+        Zip::from(self.z.view())
+            .and(self.n.view())
+            .map_collect(|z, n| {
+                calculate_weight(
+                    *z,
+                    *n,
+                    self.params.alpha(),
+                    self.params.beta(),
+                    self.params.l1_ratio(),
+                    self.params.l2_ratio(),
+                )
+            })
+    }
+
+    /// Update method of the model hyperparameters in case of async mode.
+    /// To use this method, we must store probabilities and features for the row, and update once the result (bool) is observed.
+    pub fn update<D: Data<Elem = F>, T: AsSingleTargets<Elem = bool>>(
+        &mut self,
+        dataset: &DatasetBase<ArrayBase<D, Ix2>, T>,
+        probabilities: ArrayView1<Pr>,
+    ) {
+        let gradient = calculate_gradient(probabilities.view(), dataset);
+        let sigma = self.calculate_sigma(gradient.view());
+        self.update_params(gradient, sigma);
+    }
+
+    fn predict_probabilities<D: Data<Elem = F>>(&self, x: &ArrayBase<D, Ix2>) -> Array1<Pr> {
+        let weights = self.get_weights();
+        let mut probabilities = x.dot(&weights);
+        probabilities.mapv_inplace(sigmoid);
+        probabilities.mapv(|v| Pr::new(F::to_f32(&v).unwrap_or_default()))
+    }
+
+    fn calculate_sigma(&self, gradients: ArrayView1<F>) -> Array1<F> {
+        Zip::from(&self.n)
+            .and(gradients)
+            .map_collect(|n, grad| calculate_sigma(*n, *grad, self.params.alpha))
+    }
+
+    fn update_params(&mut self, gradient: Array1<F>, sigma: Array1<F>) {
+        let weights = self.get_weights();
+        self.z += &gradient;
+        self.z -= &(sigma * weights);
+        self.n += &gradient.mapv(|grad| grad.powf(F::cast(2.)));
+    }
+}
+
+fn calculate_gradient<F: Float, D: Data<Elem = F>, T: AsSingleTargets<Elem = bool>>(
+    probabilities: ArrayView1<Pr>,
+    dataset: &DatasetBase<ArrayBase<D, Ix2>, T>,
+) -> Array1<F> {
+    let targets = dataset.as_single_targets();
+    let x = dataset.records();
+    let diff = Zip::from(&probabilities)
+        .and(targets)
+        .map_collect(|prob, y| {
+            let truth = if *y { F::one() } else { F::zero() };
+            F::cast(**prob) - truth
+        });
+    diff.dot(x)
+}
+
+fn calculate_sigma<F: Float>(n: F, gradient: F, alpha: F) -> F {
```

Comment:
this weights the current solution to the dual sum, so why not call it `calculate_weight_in_average` so that people reading the `update` function have more pointers how it relates to RDA (than just comparing 1-to-1 to the paper with sigma)